src/v/cluster/utils/notifications.h

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+#include "cluster/notification.h"
+#include "cluster/topic_configuration.h"
+#include "model/fundamental.h"
+
+#include <seastar/util/noncopyable_function.hh>
+
+namespace cluster {
+
+// Unified interface for partition level notifications.
+// Supports notifications for partition leadership changes,
+// partition management (creation, deletion), and partition properties changes.
+class notifications {
+public:
+    // bool is true if the ntp local to this shard is the leader
+    struct partition_state {
+        partition_state(
+          model::ntp ntp,
+          model::term_id term,
+          bool is_leader,
+          std::optional<cluster::topic_configuration> topic_cfg)
+          : ntp(std::move(ntp))
+          , term(term)
+          , is_leader(is_leader)
+          , topic_cfg(std::move(topic_cfg)) {}
+        model::ntp ntp;
+        model::term_id term;
```

Comment:
YAGNI? I don't know where we would care about term for these use cases...

Reply:
I started without term but perhaps need it for DR.. the rationale is term_id acts as  a simple fence in some cases.. for example we try to process a notification at term=0 but the processing doesn't succeed (we queue it again) but meanwhile there is a notification processed in term term=1 which succeeds and then the future term=0 op becomes a noop.

Reply:
Yeah I'd personally rather see added when needed, but fine to keep it...